src/transformers/trainer.py

```diff
@@ -666,3 +723,19 @@ def _prediction_loop(
             metrics["loss"] = np.mean(eval_losses)
 
         return PredictionOutput(predictions=preds, label_ids=label_ids, metrics=metrics)
+
+    def distributed_concat(self, arr: np.ndarray, num_total_examples: int) -> np.ndarray:
+        assert self.args.local_rank != -1
+
+        def create_copy():
+            return torch.from_numpy(arr).to(self.args.device)
+
+        output_tensors = [create_copy() for _ in range(torch.distributed.get_world_size())]
+        torch.distributed.all_gather(output_tensors, create_copy())
```

Comment:
You don't actually need to do all these copies of the initial tensor, you can rely on `repeat` to reallocate the Tensor to the specified number of chunks you need:

```python
world_size = torch.distributed.get_world_size()

# Create the PyTorch backed Tensor
pt_arr = torch.from_numpy(arr).to(self.args.device)

# Repeat actually reallocate at once the tensor to the specific size (i.e. number of workers)
pt_arr_repeated = pt_arr.unsqueeze(0).repeat(world_size, 1, 1).chunk(world_size)

# pt_arr_repeated[0] is the actual data we would like to send to the other
torch.distributed.all_gather(pt_arr_repeated, pt_arr_repeated[0])
```

Reply:
`all_gather` will dispatch the final tensor to every device belonging to the same `ProcessGroup` (often the entire world). I tend to agree the `gather` would be more appropriate, but I'm not sure `NCCL` actually supports it.

As we are not sending huge gradients buffers over the medium, but just few scalar values, I suppose `all_gather` is fine and will not generate congestion on the network.

Reply:
Yes that's exactly why I used all_gather instead of gather: see table at the top of https://pytorch.org/docs/stable/distributed.html